agent/php_execute.c

```diff
@@ -1126,6 +1111,17 @@ static inline void nr_php_execute_segment_end(
       || stacked->error) {
     nr_segment_t* s = nr_php_stacked_segment_move_to_heap(stacked TSRMLS_CC);
     nr_php_execute_segment_add_metric(s, metadata, create_metric);
+
+    /*
+     * Check if code level metrics are enabled in the ini.
+     * If they aren't, exit and don't add any attributes.
+     */
+    if (NRINI(code_level_metrics_enabled)) {
+      if (NULL == s->attributes) {
+        s->attributes = nr_attributes_create(s->txn->attribute_config);
```

Comment:
How likely it is that attributes will have to be created here? I ask because they potentially get created (malloc'ed) here and maybe not even used if [this condition](https://github.com/zsistla/newrelic-php-agent/blob/clm_to_dev/agent/php_txn.c#L1181) or [this condition](https://github.com/zsistla/newrelic-php-agent/blob/clm_to_dev/agent/php_txn.c#L1187) are not met. Since this will get called for every user function execution, when CLM is enabled, maybe the malloc could be postponed until it's absolutely necessary?

Reply:
I'll see if I can move it into the nr_php_txn_add_code_level_metrics function to after we do our valid clm checks.  It's very likely to need to be created but it's only called for the segments we keep not for every user function.

Reply:
Maybe consider changing `nr_php_txn_add_code_level_metrics` signature to take the segment as argument as I suggest [here](https://github.com/newrelic/newrelic-php-agent/pull/585#discussion_r1043702320). That should make it possible to move attributes allocation after all conditions to add them are met.

Reply:
Addressed in 3e7ec823504a22e68c094b947fd86af06ebd41d9.